### ISSUE TITLE
Lerna-based projects generate "JavaScript heap out of memory"

### ISSUE THREAD

**Issue by k3n** [GoogleChromeLabs/ndb#89]
### Steps to reproduce

**Tell us about your environment:**

* ndb version: 1.0.16
* Platform / OS version: macOS 10.13.4 (Darwin 17.5.0)
* Node.js version: 8.11.1

**What steps will reproduce the problem?**

Using [Lerna](https://github.com/lerna/lerna), I have a directory structure as follows:

```
.
+-- package.json
+-- node_modules
|   +-- lerna
|   +-- babel
|   +-- etc.
+-- packages
|   +-- package-a
|   |   +-- package.json
|   |   +-- node_modules
|   |   |   +-- package-b -> ../../package-b # symlinks
|   |   |   +-- package-c -> ../../package-c
|   |   |   +-- etc. (8 packages total)
|   +-- package-b
|   |   +-- package.json
|   |   +-- node_modules
|   +-- package-c
|   |   +-- package.json
|   |   +-- node_modules
|   +-- etc.
```

The server begins to start -- I see logs generated by my application -- but at some point before it completes, I get the following in the console:

```
$ ndb packages/package-a/index.js
Debugger listening on ws://127.0.0.1:57593/7546efc2-8ffb-46f9-8ba1-8b63ae8a352c
For help see https://nodejs.org/en/docs/inspector
Debugger listening on ws://127.0.0.1:57594/f759a14c-0514-492d-ac5f-460eeb29f784
For help see https://nodejs.org/en/docs/inspector
Debugger attached.
Debugger attached.

# application startup logs
# (snipped for brevity)

<--- Last few GCs --->

[23721:0x102801e00]   118155 ms: Mark-sweep 1360.4 (1426.5) -> 1360.4 (1427.5) MB, 460.5 / 0.0 ms  allocation failure GC in old space requested
[23721:0x102801e00]   118645 ms: Mark-sweep 1360.4 (1427.5) -> 1360.4 (1420.0) MB, 489.8 / 0.0 ms  last resort GC in old space requested
[23721:0x102801e00]   119122 ms: Mark-sweep 1360.4 (1420.0) -> 1360.4 (1420.0) MB, 476.6 / 0.0 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x397ff8d257c1 <JSObject>
    1: DoJoin(aka DoJoin) [native array.js:~94] [pc=0x10bd26d24e43](this=0x397f5d6022d1 <undefined>,o=0x397fc22ddb09 <JSArray[3]>,p=3,D=0x397f5d602371 <true>,z=0x397ff8d3ea79 <String[1]: ,>,y=0x397f5d6023e1 <false>)
    2: Join(aka Join) [native array.js:~119] [pc=0x10bd26d293de](this=0x397f5d6022d1 <undefined>,o=0x397fc22ddb09 <JSArray[3]>,p=3,z=0x397ff8d3ea79 <String[1]: ,>,y=0x397f5d6023e1 <f...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/Users/k3n/.nvm/versions/node/v8.11.1/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/Users/k3n/.nvm/versions/node/v8.11.1/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/Users/k3n/.nvm/versions/node/v8.11.1/bin/node]
 4: v8::internal::Factory::NewRawOneByteString(int, v8::internal::PretenureFlag) [/Users/k3n/.nvm/versions/node/v8.11.1/bin/node]
# misc. stacktrace data (changes between runs)
Abort trap: 6
$
```

_Please include code that reproduces the issue._

Code is from a private repo, sorry :(

**What is the expected result?**

I'm able to debug with `ndb`.

**What happens instead?**

I can't debug with `ndb`.

**Comment by ak239** [comment]
Could you verify that it is still a problem with ndb 1.0.20? (We landed couple fixes after 1.0.16).
Does you server produce a lot of console logs? Is it some common server (e.g. express) or written from scratch?
You can try to set breakpoint at some point of your program before crash and try to collect heap snapshot on memory tab to find what objects in your app are prevented from being collected by ndb.

**Comment by k3n** [comment]
> Could you verify that it is still a problem with ndb 1.0.20? (We landed couple fixes after 1.0.16).

I plan to when time permits!

> Does you server produce a lot of console logs?

In general, not really, though it's possible at startup time to generate a hundred or so as things initialize.

> Is it some common server (e.g. express) or written from scratch?

It's hapi.js.

**Comment by ak239** [comment]
This one should be fixed in canary channel, please try `npm install -g ndb@canary`.
